pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+	isDowngrade := compat && len(v2Rules) > 0 // 4
+
+	// The following operation will be done on a per-rule basis (or
+	// per-endpoint, assuming that each egress rule has a unique IP addr
+	// associated with the endpoint).
+	//
+	// In both the upgrade and downgrade scenario, the following happens in a
+	// specific order to guarantee that any failure at any point won't cause
+	// connectivity disruption for the endpoint. Any errors encountered do not
+	// stop the migration process because we want to ensure that we conform to
+	// either the new state or the old state, and want to avoid being
+	// in-between datapath states.
+	//   1) Copy over new routes from the old routes
+	//   2) Insert new rule
+	//   3) Delete old rule
+	//   4) Delete old routes
+	// Doing (1) & (2) before (3) & (4) allows us to essentially perform an
+	// "atomic" swap-in for the new state.
+	//
+	// (4) is attempted separately outside the main loop because we want to
+	// avoid deleting routes for endpoints that share the same table ID. We
+	// will delete the routes, if and only if, all endpoints that share the
+	// same table ID succeeded in migrating. If an endpoint failed to migrate,
+	// then any routes that reference the table ID associated with the
+	// endpoint's egress rule will be skipped. This is to prevent disrupting
+	// endpoints who relying on the old state to be in-place.
+	//
+	// If a failure occurs at (1), then the old state can continue to service
+	// the endpoint. Similarly with (2) because routes without rules are likely
+	// to not have any effect.
+	//
+	// If a failure occurs at (3), we have already succeeded in getting the new
+	// state in-place to direct traffic for the endpoint. In any case of
+	// upgrade or downgrade, it is possible for both states to be in-place if
+	// there are any failures, especially if there were any failures in
+	// reverting. The datapath selected will depend on the rule priority.
+	//
+	// Concretely, for upgrades, the newer rule will have a lower priority, so
+	// the original datapath will be selected. The migration is deemed a
+	// failure because the original datapath (with a rule that has a higher
+	// priority) is being selected for the endpoint. It is necessary to attempt
+	// reverting the failed migration work [(1) & (2)], as leaving the state
+	// could block a user's from retrying this upgrade again.
+	//
+	// For downgrades, the newer rule will have a higher priority, so the newer
+	// datapath will be selected. The migration is deemed a success and we
+	// explicitly avoid reverting, because it's not necessary to revert this
+	// work merely because we failed to cleanup old, ineffectual state.
+	//
+	// In either case, no connectivity is affected for the endpoint.
+	//
+	// If we fail at (4), then the old rule will have been deleted and the new
+	// state is in-place, which would be servicing the endpoint. The old routes
+	// would just be leftover state to be cleaned up at a later point.
```

Comment:
Is there any mechanism currently in place to clean up rules and routes left over from a partial migration "at a later point"? It may not be needed in practice of course.

Reply:
There is not. There are certain scenarios that are more risky than others if we fail to cleanup and we warn the user of those specific cases. Otherwise, we perform a best-effort to cleanup. If something fails, we'll tell the user that they should clean it up, but otherwise datapath connectivity is unaffected.